File: src/main/java/com/husu/threesum/ThreeSum.java
package com.husu.threesum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author huyong(husu)
 * @date 5/28/2024 4:59 PM
 */
public class ThreeSum {
    public static void main(String[] args) {
        int[] nums = {-1, 0, 1, 2, -1, -4};
        int target = 0;
        List<List<Integer>> result = threeSum(nums, target);
        for (List<Integer> integers : result) {
            System.out.println(integers);
        }
    }

    public static List<List<Integer>> threeSum(int[] nums, int target) {
        Arrays.sort(nums);
        int len = nums.length;
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < len; i++) {
            if (nums[i] > 0) {
                break;
            }
            if (nums[len - 1] < 0) {
                break;
            }

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = len - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == target) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;

                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right--;
                    }
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }
}

File: src/main/java/com/husu/threesum/ThreeSum.js
function threeSum(nums) {
    let results = [];
    nums.sort((a, b) => a - b);  // 将数组排序

    for (let i = 0; i < nums.length - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;  // 避免重复
        }

        let left = 0, right = nums.length - 1;
        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];
            if (sum === 0) {
                results.push([nums[i], nums[left], nums[right]]);
                left++;
                right--;
                while (left < right && nums[left] === nums[left + 1]) {
                    left++;  // 跳过重复
                }
                while (left < right && nums[right] === nums[right - 1]) {
                    right--;  // 跳过重复
                }
                left++;
                right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }

    return results;
}

// 示例
const nums = [-1, 0, 1, 2, -1, -4];
console.log(threeSum(nums));  // 输出: [[-1, -1, 2], [-1, 0, 1]]
